drivers/i2c/i2c_ite_it8xxx2.c

```diff
@@ -309,6 +323,373 @@ static void i2c_w2r_change_direction(const struct device *dev)
+		if (port == 0) {
+			IT8XXX2_SMB_I2CW2RF &= ~(IT8XXX2_SMB_MAIF |
+						 IT8XXX2_SMB_MAIFI);
+		} else if (port == 2) {
+			IT8XXX2_SMB_I2CW2RF &= ~(IT8XXX2_SMB_MBCIF |
+						 IT8XXX2_SMB_MCIFI);
+		}
+	}
+
+	irq_unlock(key);
+}
+
+static int i2c_tran_fifo_write_to_read(const struct device *dev)
+{
+	struct i2c_it8xxx2_data *data = dev->data;
+	const struct i2c_it8xxx2_config *config = dev->config;
+	uint32_t i;
+	uint8_t *base = config->base;
+	volatile uint8_t *reg_mstfctrl = config->reg_mstfctrl;
+
+	if (data->msgs->flags & I2C_MSG_START) {
+		/* Clear start flag. */
+		data->msgs->flags &= ~I2C_MSG_START;
+		/* Enable SMB channel in FIFO mode. */
+		*reg_mstfctrl |= IT8XXX2_SMB_FFEN;
+		/* Enable I2C write to read FIFO mode. */
+		i2c_fifo_en_w2r(1, config->port);
+		/* I2C enable. */
+		IT8XXX2_SMB_HOCTL2(base) = IT8XXX2_SMB_SMD_TO_EN |
+					   IT8XXX2_SMB_I2C_EN |
+					   IT8XXX2_SMB_SMHEN;
+		/* Set write byte counts. */
+		IT8XXX2_SMB_D0REG(base) = data->msgs->len;
+		/*
+		 * bit[7:1]: Address of the targeted slave.
+		 * bit[0]: Direction of the host transfer.
+		 */
+		IT8XXX2_SMB_TRASLA(base) = (uint8_t)data->addr_16bit << 1;
+		/* The maximum fifo size is 32 bytes. */
+		data->bytecnt = MIN(data->msgs->len, I2C_FIFO_MODE_MAX_SIZE);
+
+		for (i = 0; i < data->bytecnt; i++) {
+			/* Set host block data byte. */
+			IT8XXX2_SMB_HOBDB(base) = *(data->msgs->buf++);
+		}
+		/* Calculate the remaining byte counts. */
+		data->bytecnt = data->msgs->len - data->bytecnt;
+		/*
+		 * bit[6] = 1b: Start.
+		 * bit[4:2] = 111b: Extend command.
+		 * bit[0] = 1b: Host interrupt enable.
+		 */
+		IT8XXX2_SMB_HOCTL(base) = IT8XXX2_SMB_SRT |
+					  IT8XXX2_SMB_SMCD_EXTND |
+					  IT8XXX2_SMB_INTREN;
+	} else {
+		/* Check block done status. */
+		if (*reg_mstfctrl & IT8XXX2_SMB_BLKDS) {
+			if (IT8XXX2_SMB_HOCTL2(base) & IT8XXX2_SMB_I2C_SW_EN) {
+				for (i = 0; i < I2C_FIFO_MODE_MAX_SIZE; i++) {
+					/* To get received data. */
+					*(data->msgs->buf++) =
+						IT8XXX2_SMB_HOBDB(base);
+				}
+				/* Clear FIFO block done status. */
+				*reg_mstfctrl |= IT8XXX2_SMB_BLKDS;
+				/* Calculate the remaining byte counts. */
+				data->bytecnt -= I2C_FIFO_MODE_MAX_SIZE;
+			} else {
+				uint32_t _bytecnt;
+
+				/* The maximum fifo size is 32 bytes. */
+				_bytecnt = MIN(data->bytecnt,
+					       I2C_FIFO_MODE_MAX_SIZE);
+
+				for (i = 0; i < _bytecnt; i++) {
+					/* Set host block data byte. */
+					IT8XXX2_SMB_HOBDB(base) =
+						*(data->msgs->buf++);
+				}
+				/* Clear FIFO block done status. */
+				*reg_mstfctrl |= IT8XXX2_SMB_BLKDS;
+				/* Calculate the remaining byte counts. */
+				data->bytecnt -= _bytecnt;
+			}
+		} else if (IT8XXX2_SMB_IWRFISTA & BIT(config->port)) {
+			/* Set I2C_SW_EN = 1 */
+			IT8XXX2_SMB_HOCTL2(base) |=
+				IT8XXX2_SMB_I2C_SW_EN |
+				IT8XXX2_SMB_I2C_SW_WAIT;
+			IT8XXX2_SMB_HOCTL2(base) &=
+				~IT8XXX2_SMB_I2C_SW_WAIT;
+			/* Point to the next msg for the read location. */
+			data->msgs++;
```

Comment:
These seems fragile, and can possibly index past the end of the `msgs` array passed to `i2c_it8xxx2_transfer()`.  

I think you should store a copy of the `msgs` pointer and an index tracking the current message.

```
struct i2c_it8xxx2_data {
#ifdef CONFIG_I2C_IT8XXX2_FIFO_MODE
    uint32_t bytecnt;
    uint8_t num_msgs;
    uint8_t active_msg_index;
    struct i2c_msg *msgs_list;  /* Set by i2c_it8xxx2_transfer() */
#endif
```

And then here, you would do the following:
```
if (++data->active_msg_index >= data->num_msgs)
    /* Log a failure */
data->active_msg = &data->msgs_list[data->active_msg_index];
```

Reply:
Done.